parallel-exec/lib/lib.go

```diff
@@ -0,0 +1,418 @@
+	runnerOptions := newRunnerOptions()
+	for _, option := range options {
+		option(runnerOptions)
+	}
+	return &runner{runnerOptions}
+}
+
+func (r *runner) Run(cmds []*exec.Cmd) error {
+	// do not want to acquire lock in the signal handler
+	// do there is a race condition where err could be set to
+	// errCmdFailed or not set at all even after an interrupt happens
+	var err error
+	doneC := make(chan struct{})
+	cmdControllers := make([]*cmdController, len(cmds))
+	for i, cmd := range cmds {
+		cmdControllers[i] = newCmdController(cmd, r.options.EventHandler, r.options.Timer)
+	}
+
+	signalC := make(chan os.Signal, 1)
+	signal.Notify(signalC, os.Interrupt)
+	go func() {
+		for _ = range signalC {
+			err = errInterrupted
+			doneC <- struct{}{}
+			return
+		}
+	}()
+
+	var waitGroup sync.WaitGroup
+	semaphore := newSemaphore(r.options.MaxConcurrentCmds)
+
+	startTime := r.options.Timer()
+	r.options.EventHandler(newStartedEvent(startTime))
+	for _, cmdController := range cmdControllers {
+		cmdController := cmdController
+		waitGroup.Add(1)
+		go func() {
+			semaphore.P(1)
+			defer semaphore.V(1)
+			defer waitGroup.Done()
+			if !cmdController.Run() {
+				// best effort to prioritize the interrupt error
+				// but this is not deterministic
+				err = errCmdFailed
+				if r.options.FastFail {
+					doneC <- struct{}{}
+				}
+			}
+		}()
+	}
+	go func() {
+		// if everything finishes and there is an interrupt, we could
+		// end up not actually returning an error if everything below
+		// complietes before we context switch to the interrupt goroutine
+		waitGroup.Wait()
+		doneC <- struct{}{}
+	}()
+	<-doneC
+	for _, cmdController := range cmdControllers {
+		cmdController.Kill()
+	}
+	finishTime := r.options.Timer()
+	r.options.EventHandler(newFinishedEvent(finishTime, startTime, err))
+	return err
+}
+
+type cmdController struct {
+	Cmd          *exec.Cmd
+	EventHandler EventHandler
+	Timer        Timer
+	Started      bool
+	Finished     bool
+	StartTime    time.Time
+	Lock         sync.Mutex
+}
+
+func newCmdController(cmd *exec.Cmd, eventHandler EventHandler, timer Timer) *cmdController {
+	return &cmdController{cmd, eventHandler, timer, false, false, timer(), sync.Mutex{}}
+}
+
+// Run returns false on failure that has not been already handled
+func (c *cmdController) Run() bool {
+	c.Lock.Lock()
+	if c.Started || c.Finished {
+		c.Lock.Unlock()
+		return true
+	}
+	c.Started = true
+	c.StartTime = c.Timer()
+	c.EventHandler(newCmdStartedEvent(c.StartTime, c.Cmd))
+	if err := c.Cmd.Start(); err != nil {
+		finishTime := c.Timer()
+		err = fmt.Errorf("command could not start: %s: %s", cmdString(c.Cmd), err.Error())
+		c.Finished = true
+		c.EventHandler(newCmdFinishedEvent(finishTime, c.Cmd, c.StartTime, err))
+		c.Lock.Unlock()
+		return false
+	}
+	c.Lock.Unlock()
+	err := c.Cmd.Wait()
+	finishTime := c.Timer()
+	if err != nil {
+		err = fmt.Errorf("command had error: %s: %s", cmdString(c.Cmd), err.Error())
+	}
+	c.Lock.Lock()
```

Comment:
there's a lot of lock/unlock, and not much separation in this function. a few blank lines, and maybe breaking this into multiple functions will help the readability here.